### PR TITLE
Support GEOSHAPE field type in RediSearch

### LINKED ISSUE #188: Add support for GEOSHAPE index fields and WITHIN and CONTAINS query operators
[Since RediSearch 2.8, it's possible to do more advanced GEO querying with GEOSHAPE fields.](https://docs.redis.com/latest/stack/release-notes/redisearch/redisearch-2.8-release-notes/#whats-new-in-284)

On the client side, we should provide support for Index creation and Querying. APIs should allow users to use popular WKT libraries that parse and "serialise" to WKT format (e.g. https://github.com/cschwarz/wkx)

Consider the following type-script pseudo-code:
```ts
// Geometry primitives will be used from third-party libraries and serialized to WKT
let p = new wkx.Polygon([new wkx.Point(1, 2), new wkx.Point(3, 4), new wkx.Point(5, 6), new wkx.Point(1, 2)]);

await client.hSet('land:123', {
    geometry: p.toWkt(), // Serialisation
    price: 29000
})

await client.ft.create('idx:land', {
    'geometry': {
        type: SchemaFieldTypes.GEOSHAPE // New field type
        COORD_SYSTEM: GeoshapeFieldCoordSystem.FLAT // Coordinate system type 
    },
    'price': {
        type: SchemaFieldTypes.NUMERIC,
    }
}, {
    ON: 'HASH',
    PREFIX: 'land:'
});

let areaOfInterest = new wkx.Polygon([new wkx.Point(1, 2), new wkx.Point(3, 4), new wkx.Point(5, 6), new wkx.Point(1, 2)]);

let result = await client.ft.search(
    'idx:land',
    '@geometry:[WITHIN $area] @price:[20000 30000]',
    {
        PARAMS: {
            area: areaOfInterest.toWkt()
        }
    }
);

```

Requirements:
- Add support for GEOSHAPE fields:
  - Perform validations to allow passing only `POLYGON` and `POINT` geometry primitives.
  - Support specifying the coordinate system
- Ensure that querying API is compatible with `GEOSHAPE` queries 

Reference documentation:
https://redis.io/commands/ft.create/
https://redis.io/commands/ft.search/

### PR DESCRIPTION
**Pull request by shacharPash**
Closes #188

### CODE REVIEW

File: Examples/GeoShape.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
+# GeoShape Fields Usage In RediSearch
+
+As of RediSearch 2.8.4, advanced GEO querying with GEOSHAPE fields is supported.
+
+Any object/library producing a
+[well-known text (WKT)](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry) in `string` format can be used.
+
+In this example, we'll demonstrate how to use GeoShape fields in RediSearch with [NetTopologySuite](https://github.com/NetTopologySuite/NetTopologySuite) library.
```
Comment (chayim) [inline Examples/GeoShape.md:8]: `Using GeoShape fields in searches with the [NetTopologySuite](https://github.com/NetTopologySuite/NetTopologySuite) library.`

File: src/NRedisStack/NRedisStack.csproj [hunk 2]
```diff
@@ -15,8 +15,9 @@
   </PropertyGroup>
 
   <ItemGroup>
+	<PackageReference Include="NetTopologySuite" Version="2.5.0" />
 	<PackageReference Include="System.Text.Json" Version="7.0.2" Condition="'$(TargetFramework)' == 'netstandard2.0'" />
-	<PackageReference Include="StackExchange.Redis" Version="2.6.96" />
+	<PackageReference Include="StackExchange.Redis" Version="2.6.122" />
```
Comment (chayim) [inline src/NRedisStack/NRedisStack.csproj:20]: Let's note in the PR that we're upgrading the version of SE.Redis. Ideally this belongs in a separate PR even.

Comment (shacharPash) [reply src/NRedisStack/NRedisStack.csproj:20]: I saw that there are projects that do it automatically, maybe we should do it for SE.Redis as well

File: Examples/GeoShape.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
+# GeoShape Fields Usage In RediSearch
+
+NRedisStack now supports GEOSHAPE field querying.
+
+
```
Comment (chayim) [inline Examples/GeoShape.md:5]: extra whitespace

File: Examples/GeoShape.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
+# GeoShape Fields Usage In RediSearch
```
Comment (chayim) [inline Examples/GeoShape.md:1]: Can we rename the file GeoShapeQueryExample.md? Our other examples follow a similar pattern.

File: Examples/GeoShape.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+The search result from redis is:
+
+```bash
+1) (integer) 1
+2) "small"
+3) 1) "geom"
+   2) "POLYGON ((1 1, 1 100, 100 100, 100 1, 1 1))"
+```
+
+we can use the reader to get the polygon object:
```
Comment (chayim) [inline Examples/GeoShape.md:75]: `Use the reader to get the polygon:`

File: Examples/GeoShape.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+### Querying within condition
+
+```csharp
+Polygon within = factory.CreatePolygon(new Coordinate[]{new Coordinate(0, 0),
+new Coordinate(0, 150), new Coordinate(150, 150), new Coordinate(150, 0), new Coordinate(0, 0)});
+
+SearchResult res = ft.Search(index, new Query("@geom:[within $poly]").AddParam("poly", within.ToString()).Dialect(3));
+```
+
+The search result from redis is:
```
Comment (chayim) [inline Examples/GeoShape.md:66]: `Our search result`:

File: Examples/GeoShape.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+```csharp
+Polygon contains = factory.CreatePolygon(new Coordinate[]{new Coordinate(2, 2),
+new Coordinate(2, 50), new Coordinate(50, 50), new Coordinate(50, 2), new Coordinate(2, 2)});
+
+res = ft.Search(index, new Query("@geom:[contains $poly]").AddParam("poly", contains.ToString()).Dialect(3)); // DIALECT 3 is required for this query
+
+```
+
+The search result from redis is:
```
Comment (chayim) [inline Examples/GeoShape.md:91]: `Our search result`

File: Examples/GeoShape.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+4) "large"
+5) 1) "geom"
+   2) "POLYGON ((1 1, 1 200, 200 200, 200 1, 1 1))"
+```
+
+### Point type
+
+```csharp
+Point point = factory.CreatePoint(new Coordinate(10, 10));
+db.HashSet("point", "geom", point.ToString());
```
Comment (chayim) [inline Examples/GeoShape.md:107]: Add a comment maybe - or something like this.

# Note serializing the argument to string

File: src/NRedisStack/Search/Schema.cs [hunk 9]
```diff
@@ -38,6 +40,7 @@ internal void AddSchemaArgs(List<object> args)
                 {
                     FieldType.Text => "TEXT",
                     FieldType.Geo => "GEO",
+                    FieldType.GeoShape => "GEOSHAPE",
```
Comment (chayim) [inline src/NRedisStack/Search/Schema.cs:43]: Good

**Review by chayim — APPROVED** [review]
